sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+used at a time depending on whether 3rd-party host compiler is used or not.
+
+Integration footer can be used with 3rd-party host compilers, but it requires
+appending to a translation unit provided by user, which could affect debug
+information: since there are no compilers that support appending a file at the
+end (similar to `-include`), appending is done by generating a temporary input
+file using concatenation of the original input and integration footer.
+
+Such replacement of the main translation unit causes the following issues:
+- debug information about the source file might be incorrect, leading to
+  problems with gdb `l` command and code coverage tools
+- checksum of host and device source files becomes different which causes device
+  code debugging to be completely broken in some environments (such as MS Visual
+  Studio, for example)
+
+Customizing host compiler allows to avoid issues with debuggers and code
+coverage tools, but that is not an option if user wants to compile host part
+of an app with a 3rd-party host compiler.
+
+Further sections describe the implementation design of both approaches in more
+details, note that there are few components which should be modified regardless
+of which approach is in use.
+
+## Common front-end part
+
+DPC++ FE should support the following attribute:
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]`. This attribute accepts
+a string literal and should be applied to types (like `device_global` or
+`specialization_id`).
+
+Presence of the attribute instructs the compiler to perform the following
+things:
+- emit `sycl-unique-id` LLVM IR attribute on each definition of a variable of
+  type marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]]`
+  attribute. `sycl-unique-id` LLVM IR attribute should be accompanied by a
+  unique string identifier of a variable it is attached to. The rules for
+  creating this string are the same as for `__builtin_sycl_unique_stable_id` and
+  the same algorithm can be used when generating the string for the attribute
+- emit `sycl-uid-type` LLVM IR attribute alongside `sycl-unique-id`, which
+  contains the `kind` string passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute
+
+**TODO**: we have `[[__sycl_detail__::device_global]]` attribute documented in
+[device global design doc][device-global-design], which instructs front-end to
+emit some additional semantic checking. Shall we leave it in place or that
+request for semantic checking should also be documented by
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute when `kind`
+is set to a certain value?
+
+**TODO**: alternatively, we could completely re-use existing
+`[[__sycl_detail__::device_global]]` attribute and introduce another one for
+specialization constants, i.e. it is a question of whether or not we want to
+generalize unique IDs generation in form of a generic attribute or not.
```

Comment:
I think I have a slight preference to keep `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` and `[[__sycl_detail__::device_global]]` as separate attributes.

FWIW, I originally proposed a general attribute like `[[__sycl_detail__::restrictions("device_global")]]`, where the parameter would tell the CFE which set of restrictions to enforce.  However, the CFE team preferred having a separate C++ attribute for each set of restriction.  Thus, we have `[[__sycl_detail__::device_global]]` instead.

Reply:
So, having `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` for setting LLVM IR attribute and having separate `[[__sycl_detail__::device_global]]` for semantic checks is the preferable way, right?

> In order to implement such attribute we should hardcode and keep track of possible "kinds" of uniquely identifiable objects in FE, meaning the FE should diagnose each time the passed kind is not valid and it should be updated for each new kind.

Note that for LLVM IR attribute generation purposes I do not suggest that FE perform any correctness check for `kind` argument except that it is non empty: FE here just serves as a translation layer and headers are responsible to pass the correct string there to ensure that everything will work fine.

Reply:
> So, having `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` for setting LLVM IR attribute and having separate `[[__sycl_detail__::device_global]]` for semantic checks is the preferable way, right?

That sounds good to  me.

Reply:
> Note that for LLVM IR attribute generation purposes I do not suggest that FE perform any correctness check for kind argument except that it is non empty: FE here just serves as a translation layer and headers are responsible to pass the correct string there to ensure that everything will work fine.

Doesn't that introduce the dependency between the headers that will use the attribute and the compiler's layers which will use the "kind" string? I guess somehow it should be made sure that the string is correct, not issuing a diagnostic from FE will make it up to developers.

> So, having [[__sycl_detail__::uniquely_identifiable_object(kind)]] for setting LLVM IR attribute and having separate [[__sycl_detail__::device_global]] for semantic checks is the preferable way, right?

In my mind it is preferable to leave `[[__sycl_detail__::device_global]]` "as is" and introduce another attribute not accepting any arguments for specialization ids that also may enforce some restrictions and will trigger emission of LLVM IR attributes. No magic strings.

Reply:
> Doesn't that introduce the dependency between the headers that will use the attribute and the compiler's layers which will use the "kind" string? I guess somehow it should be made sure that the string is correct, not issuing a diagnostic from FE will make it up to developers.

I propose that `kind` string is not parsed by the compiler and simply propagated as-is to runtime through `__register_uniquely_identifiable_object`. That means that there is no dependency between the compiler and the runtime, because runtime should simply expect the same tag it passed to the attribute in headers. 

Various semantic checks/detection of certain variables for additional processing should happen based on some other markup. For example, there is no need to mark `specialization_id` variables to handle them properly: the compiler looks for calls to `__sycl_getSpecializationConstantValue` pseudo-intrinsics. For `device_global`, the compiler checks for `sycl-device-global-size` metadata.

To summarize:
- use existing `[[__sycl_detail__::device_global]]` for semantic checking and special markup (as it is now)
- use new `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` with `kind` not being checked/specially handled at _any_ compilation stage: this is simply a marker which set in headers and propagated as-is back to runtime, i.e. it is headers-runtime interface

Does it look ok?

Note: we could even get rid of `kind` completely, but I think that runtime will benefit from its presence. Would be good to hear from RT folks (@steffenlarsen, @s-kanaev, @romanovvlad)

Reply:
If there will be no dependency between the compiler and runtime, then it sounds ok to me.

> use existing [[__sycl_detail__::device_global]] for semantic checking and special markup (as it is now)
use new [[__sycl_detail__::uniquely_identifiable_object(kind)]] with kind not being checked/specially handled at any compilation stage: this is simply a marker which set in headers and propagated as-is back to runtime, i.e. it is headers-runtime interface

So, did I get it right that `[[__sycl_detail__::device_global]]` is left for semantic checks, but both `device_global` and `specialization_id` types are marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]` in order to emit LLVM IR attributes?

Reply:
>> Note: we could even get rid of kind completely, but I think that runtime will benefit from its presence. Would be good to hear from RT folks.

Do not see how it could be used other than for debugging in runtime. I would suggest having a `long int` argument which would be reserved for future use instead.

Reply:
It seems to me that the `kind` parameter would be useful.  Without it, the runtime won't be able to tell the difference between different calls to `__register_uniquely_identifiable_object()`.  Won't the runtime want to know whether the call is registering a "device global" variable vs. a "specialization id" variable?  It seems like the runtime might want to store these in different `std::map` objects, for example.  The runtime can only do this if it has the `kind` parameter.

Reply:
>So, did I get it right that [[__sycl_detail__::device_global]] is left for semantic checks, but both device_global and specialization_id types are marked with [[__sycl_detail__::uniquely_identifiable_object(kind)] in order to emit LLVM IR attributes?

This is my understanding as well. Can you confirm @AlexeySachkov ? This would mean changing current CodeGen behavior of` [[__sycl_detail__::device_global]]`

Reply:
>> Won't the runtime want to know whether the call is registering a "device global" variable vs. a "specialization id" variable?

The type should be known in all places where runtime needs to map an object address to a string, no?

>>  It seems like the runtime might want to store these in different std::map objects, for example.

I believe a simple `std::unordered_map` with a an address(which should be unique) of the object as a key should be enough.

Reply:
> > Won't the runtime want to know whether the call is registering a "device global" variable vs. a "specialization id" variable?
>
>The type should be known in all places where runtime needs to map an object address to a string, no?

@romanovvlad, just to understand the idea here better: you plan to have some "string" -> properties mapping available and then each time you need to do something with a spec constant or device global the runtime  will go through "address" -> "string" map and then through "string" -> properties map, right?

> > So, did I get it right that [[sycl_detail::device_global]] is left for semantic checks, but both device_global and specialization_id types are marked with [[sycl_detail::uniquely_identifiable_object(kind)] in order to emit LLVM IR attributes?
>
> This is my understanding as well. Can you confirm @AlexeySachkov ? This would mean changing current CodeGen behavior of [[__sycl_detail__::device_global]]

@Fznamznon, @elizabethandrews, yes, that is correct. `[[__sycl_detail__::device_global]]` is only supposed to be used for semantics checks and new attribute `[[sycl_detail::uniquely_identifiable_object(kind)]` will be used for attribute generation

Reply:
> The type should be known in all places where runtime needs to map an object address to a string, no?

Yes, this is true, but that's not the point I was trying to make.  Consider what the code will look like if we do *not* have the `kind` string:

```
std::unordered_map<void *, std::string> IDMap;

void __register_uniquely_identifiable_object(void *Address, const char *UniqueID) {
  IDMap[Address] = UniqueID;
}

void setSpecializationConstant(void *SpecID, /*...*/) {
  std::string UniqueID = IDMap[SpecID];  // Search through all entries in IDMap
  // ...
}

void setDeviceGlobal(void *DeviceGlobal, /*...*/) {
  std::string UniqueID = IDMap[DeviceGlobal];  // Search through all entries in IDMap
  // ...
}
```

The two  search operations are potentially more expensive because `IDMap` contain a mixture of both  "spec constant" and "device global" entries.  By contrast, consider the code if we do have `kind`:

```
std::unordered_map<void *, std::string> IDMapSpecConstants;
std::unordered_map<void *, std::string> IDMapDeviceGlobals;

void __register_uniquely_identifiable_object(void *Address, const char *UniqueID, const char *Kind) {
  if (std::string("specialization_id") == Kind) {
    IDMapSpecConstants[Address] = UniqueID;
  }
  else if (std::string("device global") == Kind) {
    IDMapDeviceGlobalks[Address] = UniqueID;
  }
}

void setSpecializationConstant(void *SpecID, /*...*/) {
  std::string UniqueID = IDMapSpecConstants[SpecID];
  // ...
}

void setDeviceGlobal(void *DeviceGlobal, /*...*/) {
  std::string UniqueID = IDMapDeviceGlobals[DeviceGlobal];
  // ...
}
```

By keeping separate maps, the search operations are potentially faster because we only search the set that we care about.

I agree that the code works either way, but it seems like you could gain some efficiency if we know the "kind" of the ID in the registration function.

Reply:
>> The two search operations are potentially more expensive because IDMap contain a mixture of both "spec constant" and "device global" entries. 
>> By keeping separate maps, the search operations are potentially faster because we only search the set that we care about.

Do not see any benefit, AFAIK look up complexity for  `std::unordered_map`  is constant. Please, note, the pointer as a key should grantee that there no collision may happen.

Reply:
>> @romanovvlad, just to understand the idea here better: you plan to have some "string" -> properties mapping available and then each time you need to do something with a spec constant or device global the runtime will go through "address" -> "string" map and then through "string" -> properties map, right?

Yes, as a first implementation, then we can consider "merging" these maps at runtime to have only one look up.

Reply:
I agree with Vlad that I don't necessarily think the `kind` is strictly needed. The implementation would likely do something along the lines of what we currently do with `device_global` (or at least do soon) i.e. we put all entries with information about the `device_globals` in a map with a pointer key and then we have another map with string keys (the unique ID) pointing to the entries in the first map.

If we were to make a shared collection holding these common entries, I imagine we would have it be these maps with opaque pointers. When first set they would have some common "uninitialized" value and once an operation tries to get it as some specific kind (`device_global`, `specialization_id`, etc.) it allocates the entry for it.
For `device_global` and `specialization_id` this initialization would happen when loading binaries, which currently for `device_global` happens before the registration, so there are no potential race conditions. However, it is not impossible that some future object using this wouldn't be initialized until some later operation happens, in which case we could have a race condition on the initialization of the map entry.

All this said, I do think not having the kind at the time of initialization potentially adds some complexity to the implementation, as accesses may have to check if their entry exists already and may have to initialize if it isn't. Knowing through the kind during registration means that operations wouldn't have to consider the possibility of an "uninitialized" entry and could simply just cast their pointer (in the hopes that there wasn't a bug somewhere.)

I think both approaches would work, however if we choose to go with the one that passes a kind, I think we should do it through and enum rather than a string. Since CFE passes the value through without caring what it actually is, only the SYCL headers and the SYCL runtime need to agree on this value, so an `enum` would be cheaper.

> Do not see any benefit, AFAIK look up complexity for `std::unordered_map` is constant. Please, note, the pointer as a key should guarantee that there no collision may happen.

Quick note about this. I think you're right about the pointer collision, but for objects like `device_global` the string map is the more frequently used one, which has higher chance of collision the bigger the map is, so if they were to share map with others there could be some negative effects on lookup-performance, though exactly how much is hard to tell.

Reply:
>> All this said, I do think not having the kind at the time of initialization potentially adds some complexity to the implementation, as accesses may have to check if their entry exists already and may have to initialize if it isn't.
>>  Knowing through the kind during registration means that operations wouldn't have to consider the possibility of an "uninitialized" entry and could simply just cast their pointer (in the hopes that there wasn't a bug somewhere.)

Could you please clarify how kind helps to understand an "uninitialized" entry?

>> Quick note about this. I think you're right about the pointer collision, but for objects like device_global the string map is the more frequently used one, which has higher chance of collision the bigger the map is, so if they were to share map with others there could be some negative effects on lookup-performance, though exactly how much is hard to tell.

Could you please elaborate more on this? I'm thinking of
`std::unordered_map</*Object address*/void *, /*Object name*/char *>`. Which scenario would lead to a collision?

>> I think both approaches would work, however if we choose to go with the one that passes a kind, I think we should do it through and enum rather than a string. Since CFE passes the value through without caring what it actually is, only the SYCL headers and the SYCL runtime need to agree on this value, so an enum would be cheaper

+1

Reply:
> Could you please clarify how kind helps to understand an "uninitialized" entry?

The idea would be to have it create an entry with room for the information needed for the given kind, that is the allocations can be made at registration time rather than some time during execution.

> Could you please elaborate more on this? I'm thinking of `std::unordered_map</*Object address*/void *, /*Object name*/char *>`. Which scenario would lead to a collision?

So you're thinking the only shared map is the "ptr -> unique ID" which will be set by the registration, and then the object-specific operations will initialize any other maps they may need? If that's the case, I agree that would be a good solution.

Reply:
Ok, it seems like we don't have an immediate need for `kind` and therefore, to simplify the design, I'm going to remove that argument - please let me know if you have any objections

Reply:
@AlexeySachkov 
>> please let me know if you have any objections

Can we have a long int argument for future use? We can have on RT level only as a starting point - `__register_uniquely_identifiable_object` takes long int 0 as the last argument.

Reply:
> > > please let me know if you have any objections
> 
> Can we have a long int argument for future use? We can have on RT level only as a starting point - `__register_uniquely_identifiable_object` takes long int 0 as the last argument.

Ok, that works for me. I will then update the doc:
- the runtime registration function _has_ the extra long int (enum) argument
- the new attribute _does not_ have the argument

This way, our FE and middle-end implementations will be simpler, but our runtime will be ready for future potential change making it easier to maintain the same ABI